### PR TITLE
[AMM] fix(swap.rs): harden swap math and enforce post-swap k invariant

### LINKED ISSUE #5: Enforce post-swap k invariant in AMM program
The work was already done here https://github.com/logos-blockchain/logos-execution-zone/pull/373

We need to move this PR over.

### PR DESCRIPTION
**Pull request by 3esmit**
## 🎯 Purpose
Fixes #5.
Migrate the issue-scoped post-swap invariant hardening from `logos-execution-zone` [PR 373](https://github.com/logos-blockchain/logos-execution-zone/pull/373) into `lez-programs`, adapted to this repository's current AMM layout.
- Reject zero-input swaps and invalid zero-reserve pools before swap math runs.
- Replace unchecked swap math and reserve updates with checked arithmetic so arithmetic hazards fail closed.
- Enforce `new_k >= old_k` with widened multiplication and add focused AMM regression coverage.

## ⚙️ What Changed
- [x] Add early `swap_amount_in > 0` and `reserve_a != 0` / `reserve_b != 0` guards in `amm/src/swap.rs`.
- [x] Replace unchecked withdraw numerator/denominator math and reserve updates with checked operations and explicit failure messages.
- [x] Add a local `mul_u128_wide` helper backed by `u128::carrying_mul` so old and new `k` can be compared without `u128` overflow.
- [x] Add AMM unit-test helpers and regression cases for zero input, zero reserves, numerator overflow, denominator overflow, and the widened-`k` boundary.
- [x] Keep the migration scoped to `amm/src/swap.rs` and `amm/src/tests.rs`; no guest, core, or IDL changes are required.

## 🔎 Migration Notes
- Ports only the final issue-5 behavior from old PR 373, not the unrelated sync/recover/minimum-liquidity work that also lived on that old branch.
- The original upstream PR was carrying commits from other PR work as well, and those extra changes were intentionally not reintroduced here.
- The diff against this repository's `main` is scoped to 2 files in the AMM crate with 175 insertions and 4 deletions.
- The `Swap` instruction interface stays unchanged, so there is no IDL regeneration or guest-entrypoint update in this PR.

## 🔎 Review Focus
- Swap now fails before any pricing math starts for zero-input and zero-reserve states.
- Checked reserve updates and full-width `k` comparison prevent silent arithmetic wraparound in both the withdraw math and the post-swap invariant check.
- The added tests cover both the new rejection paths and the large-value boundary where `reserve_a * reserve_b` exceeds `u128`.

## 🧪 How to Test
1. `cargo check -p amm_program`
2. `RISC0_DEV_MODE=1 cargo test -p amm_program --lib`
3. `RISC0_DEV_MODE=1 cargo test -p integration_tests amm_swap`

## 🔗 Dependencies
None.

## 📋 PR Completion Checklist
- [x] Port the issue-5 swap hardening from old PR 373 into `lez-programs`
- [x] Add regression coverage for the new failure and boundary cases
- [x] Leave public AMM interfaces unchanged

### CODE REVIEW

**Comment by 3esmit** [comment]
Rebased this branch onto current `main` and resolved the swap changes against the current helper-based layout.

The post-swap invariant check was added in `amm/src/swap.rs` inside `create_swap_post_states`, right where the new pool reserves are written back to the pool definition. Since both `swap_exact_input` and `swap_exact_output` go through that helper, the `new_k >= old_k` check now applies to both swap paths.

This changes will be available in a new PR, since I was instructed to move branches to this repository instead of using my own repository.

File: amm/src/swap.rs [hunk 1]
```diff
@@ -94,11 +98,33 @@ pub fn swap(
...
+    let new_reserve_a = old_reserve_a
+        .checked_add(deposit_a)
+        .expect("Reserve A overflow on swap deposit")
+        .checked_sub(withdraw_a)
+        .expect("Reserve A underflow on swap withdrawal");
+    let new_reserve_b = old_reserve_b
+        .checked_add(deposit_b)
+        .expect("Reserve B overflow on swap deposit")
+        .checked_sub(withdraw_b)
+        .expect("Reserve B underflow on swap withdrawal");
```
[the remark this answers is not in the document]
Comment (3esmit) [reply amm/src/swap.rs:113]: You're right that those reserve updates are redundant with the guarantees already established by `swap_logic`. I ended up taking the smaller refactor instead: the widened `k` check now lives in the shared post-state helper, so both swap paths go through one place.

**Comment by 3esmit** [comment]
Replaced by #45
